### PR TITLE
lp-gateway-queue: Ensure messages are processed in order

### PR DESCRIPTION
**Pull request by cdamian**
Addressing the comment regarding LP gateway queue message processing:
https://github.com/centrifuge/centrifuge-chain/pull/1991#discussion_r1766659178

### CODE REVIEW

File: pallets/liquidity-pools-gateway-queue/src/lib.rs [hunk 1]
```diff
@@ -231,10 +259,12 @@ pub mod pallet {
 
 				weight_used.saturating_accrue(weight);
 
-				MessageQueue::<T>::remove(nonce);
+				MessageQueue::<T>::remove(last_processed_nonce);
 
 				// 1 write for removing the message
 				weight_used.saturating_accrue(T::DbWeight::get().writes(1));
```
Comment (wischli) [inline pallets/liquidity-pools-gateway-queue/src/lib.rs:265]: Nit: We need to add another write: `MessageQueue::remove` + `LastProcessedNonce::set`.

File: pallets/liquidity-pools-gateway-queue/src/lib.rs [hunk 2]
```diff
@@ -200,17 +208,34 @@ pub mod pallet {
...
 
 		fn service_message_queue(max_weight: Weight) -> Weight {
-			let mut weight_used = Weight::zero();
+			let mut last_processed_nonce = LastProcessedNonce::<T>::get();
+
+			// 1 read for the last processed nonce
+			let mut weight_used = T::DbWeight::get().reads(1);
+
+			loop {
+				if let Err(_) = last_processed_nonce.ensure_add_assign(One::one()) {
```
Comment (wischli) [inline pallets/liquidity-pools-gateway-queue/src/lib.rs:217]: Nit: 
```suggestion
			loop {
				if last_processed_nonce.ensure_add_assign(One::one()).is_err() {
```

File: pallets/liquidity-pools-gateway-queue/src/lib.rs [hunk 3]
```diff
@@ -93,6 +98,9 @@ pub mod pallet {
 			message: T::Message,
 			error: DispatchError,
 		},
+
+		/// Maximum number of messages was reached.
+		MaxNumberOfMessagesWasReached,
```
Comment (wischli) [inline pallets/liquidity-pools-gateway-queue/src/lib.rs:103]: Nit: Could add the current last processed nonce to the event payload for more info

**Review by wischli — COMMENTED** [review]
Seems to be an appropriate solution to the comment

File: pallets/liquidity-pools-gateway-queue/src/lib.rs [hunk 4]
```diff
@@ -200,17 +208,34 @@ pub mod pallet {
...
+				}
 
+				// 1 read for the message
 				weight_used.saturating_accrue(T::DbWeight::get().reads(1));
 
+				let message = match MessageQueue::<T>::get(last_processed_nonce) {
+					Some(msg) => msg,
+					// No message found, we can stop.
+					None => break,
+				};
```
Comment (lemunozm) [inline pallets/liquidity-pools-gateway-queue/src/lib.rs:237]: I think the idea is fine! But here we could have an undesired behavior:

A message could be processed with priority through `process_message()` and consequently be extracted from the `MessageQueue`. If that happens, then `last_processed_nonce` could not be found, but that doesn't mean that we want to end up iterating the message queue. TLDR: Could be holes in the `MessageQueue` storage.

We should somehow jump over the holes and continue iterating until reaching the maximum nonce added to the `MessageQueue`

Comment (lemunozm) [reply pallets/liquidity-pools-gateway-queue/src/lib.rs:237]: Maybe another storage value called `LastKnownNonce` or similar does the trick

Comment (cdamian) [reply pallets/liquidity-pools-gateway-queue/src/lib.rs:237]: `MessageNonceStore` contains the last inserted nonce, I will adapt this as per your comment, I forgot about `process_message` when tackling this. Thank you!

Comment (lemunozm) [reply pallets/liquidity-pools-gateway-queue/src/lib.rs:237]: This is one of the whys why having that method increase complexity 😅

File: runtime/integration-tests/src/cases/lp/utils.rs [hunk 5]
```diff
@@ -86,15 +86,14 @@ pub fn pool_c_tranche_1_id<T: Runtime>() -> TrancheId {
...
-	let (_tx, status, receipt) = pallet_ethereum::Pending::<T>::get()
+	let (_tx, status, _receipt) = pallet_ethereum::Pending::<T>::get()
 		.last()
 		.expect("Queue triggered evm tx.")
 		.clone();
 
 	// The sender is the sender account on the gateway
 	assert_eq!(T::Sender::get().h160(), status.from);
 	assert_eq!(status.to.unwrap().0, to.0);
-	assert!(!receipt_ok(receipt));
```
Comment (cdamian) [inline runtime/integration-tests/src/cases/lp/utils.rs:97]: Note - given that the message processor implementation of the LP gateway is transaction, this check is no longer required.

File: pallets/liquidity-pools-gateway-queue/src/lib.rs [hunk 6]
```diff
@@ -200,17 +210,43 @@ pub mod pallet {
...
+				let message = match MessageQueue::<T>::get(last_processed_nonce) {
+					Some(msg) => msg,
+					// No message found at this nonce, we can skip it.
+					None => {
+						LastProcessedNonce::<T>::set(last_processed_nonce);
+
+						// 1 write for setting the last processed nonce
+						weight_used.saturating_accrue(T::DbWeight::get().writes(1));
+
+						continue;
```
Comment (lemunozm) [inline pallets/liquidity-pools-gateway-queue/src/lib.rs:246]: Nice! I think this fixes the issue of having "holes" in the queue.

**Review by lemunozm — APPROVED** [review]
Thanks Cosmin, I think now it works fine in all scenarios!

**Pull request merged**